src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,333 @@
+
+## Tags
+
+additive character
+-/
+
+/-!
+### Definitions related to and results on additive characters
+-/
+
+section add_char_def
+
+universes u v
+
+-- The domain of our additive characters
+variables (R : Type u) [comm_ring R]
+-- The target
+variables (R' : Type v) [comm_ring R']
+
+/-- Define `add_char R R'` as `(multiplicative R) →* R'`. -/
+abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+
+end add_char_def
+
+
+section additive
+
+namespace add_char
+
+open multiplicative
+
+universes u v
+
+-- The domain of our additive characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial additive character; it sends everything to `1`. -/
+@[simps] protected
+def trivial : add_char R R' :=
+{ to_fun := λ x, 1,
+  map_one' := rfl,
+  map_mul' := λ x y, (mul_one 1).symm }
+
+/-- An additive character is *nontrivial* if it takes a value `≠ 1`. -/
+def is_nontrivial (ψ : add_char R R') : Prop := ∃ (a : R), ψ a ≠ 1
+
+/-- An additive character is nontrivial iff it is not the trivial character. -/
+lemma is_nontrivial_iff_ne_trivial (ψ : add_char R R') :
+  is_nontrivial ψ ↔ ψ ≠ add_char.trivial :=
+begin
+  refine ⟨λ ⟨a, ha⟩ hf, ne_of_eq_of_ne (trivial_apply a).symm
+                         (ne_of_eq_of_ne (congr_fun (congr_arg coe_fn hf) a).symm ha) rfl, _⟩,
+  contrapose!,
+  intro h,
+  simp only [is_nontrivial, not_exists_not] at h,
+  ext,
+  exact (h x).trans (trivial_apply x).symm,
+end
+
+/-- Define the multiplicative shift of an additive character.
+This satisfies `mul_shift ψ a x = ψ (a * x)`. -/
+def mul_shift (ψ : add_char R R') (a : R) : add_char R R' :=
+ψ.comp (add_monoid_hom.mul_left a).to_multiplicative
+
+lemma mul_shift_spec (ψ : add_char R R') (a x : R) : mul_shift ψ a x = ψ (of_add (a * x)) := rfl
+
+/-- If `n` is a natural number, then `mul_shift ψ n x = (ψ x) ^ n`. -/
+lemma mul_shift_spec' (ψ : add_char R R') (n : ℕ) (x : R) : mul_shift ψ n x = (ψ x) ^ n :=
+begin
+  rw [mul_shift_spec, ← nsmul_eq_mul, of_add_nsmul, map_pow],
+  refl,
+end
+
+/-- The product of `mul_shift ψ a` and `mul_shift ψ b` is `mul_shift ψ (a + b)`. -/
+lemma mul_shift_mul (ψ : add_char R R') (a b : R) :
+  mul_shift ψ a * mul_shift ψ b = mul_shift ψ (a + b) :=
+begin
+  erw [← monoid_hom.comp_mul ψ],
+  congr,
+  have h : ∀ (f g : R →+ R),
+           f.to_multiplicative * g.to_multiplicative = (f + g).to_multiplicative := λ f g, rfl,
+  rw h,
+  congr,
+  ext,
+  simp only [add_monoid_hom.add_apply, add_monoid_hom.coe_mul_left, add_mul],
+end
+
+/-- `mul_shift ψ 0` is the trivial character. -/
+@[simp]
+lemma mul_shift_zero (ψ : add_char R R') : mul_shift ψ 0 = add_char.trivial :=
+begin
+  ext,
+  simp only [mul_shift, add_char.trivial, monoid_hom.coe_comp, function.comp_app,
+             add_monoid_hom.to_multiplicative_apply_apply, add_monoid_hom.coe_mul_left,
+             zero_mul, of_add_zero, map_one, monoid_hom.coe_mk],
```

Comment:
```suggestion
  simp only [mul_shift_apply, zero_mul, of_add_zero, map_one, trivial_apply],
```

Reply:
I get an error "unknown identifier 'mul_shift_apply'" here, but this is fixed by adding `@[simps]` to the definition of `mul_shift`.